Add the initial version of the template. (#1)
Semmle/lgtm-enterprise-google-cloud: instructions.md

# Installing LGTM Enterprise on Google Cloud
## Preparation

> Throughout this setup guide a few variables are used that are useful to set in advance. Set the Google Cloud project you want to deploy to, the zone to create any zonal resources in and a unique name for this deployment. If you are upgrading an existing installation, set these values to the same values used when LGTM was first set up.

Reviewer: I didn't realise this meant you entered the following in the shell. I thought this was in a file somewhere. To clarify (for folks like me) add a new para:

Using the Cloud Shell command line, set values for variables called `project`, `deployment`, and `zone`. For example:

> Throughout this setup guide a few variables are used that are useful to set in advance. Set the Google Cloud project you want to deploy to, the zone to create any zonal resources in and a unique name for this deployment. If you are upgrading an existing installation, set these values to the same values used when LGTM was first set up.

Reviewer: ```suggestion
Throughout this setup guide a few variables are used that are useful to set in advance. Specify an existing Google Cloud project that you want to deploy to, the zone to create any zonal resources in and a unique name of your choice for this deployment. If you are upgrading an existing installation, set these values to the same values used when LGTM was first set up.
```

```console
project="my-project"

> deployment="lgtm"
> zone="us-central1-a"

Reviewer: Swap to match the same order as referred to in the details above (not necessary, I know, I just think it reads nicer to keep things in the same order):
```suggestion
zone="us-central1-a"
deployment="lgtm"
```

```

> 

Reviewer: It wasn't immediately obvious that this panel had multiple pages. To make this clearer add:

After you have done this, click **Start** to find out how to either create or upgrade a control pool machine (step 1), or create or upgrade worker host machines (step 2).

_Note: could we use "control pool" rather than "controller"? The reason for asking is that that's the terminology we agreed with Julian way back and is used throughout the help and the Install Guide._

Author: > *Note: could we use "control pool" rather than "controller"? The reason for asking is that that's the terminology we agreed with Julian way back and is used throughout the help and the Install Guide.*

I'm happy to change it, but it makes less sense with image deployment as there is only ever one controller machine.

Reviewer 2: yes, that's true. I'm not particularly happy about the "pool" terminology (control pool, work pool, web pool) as I'm not sure that vision ever became a reality at many customer sites. But that's what we've got across lots of documentation. However, I've checked back in the Architecture Guide and we do mention "controller" also (if briefly), so let's stick with that.

i.e. please ignore my "controller" > "control pool" suggestsions.
Thanks

> ## Controller

Reviewer: ```suggestion
## Control pool
```

> ### Creating a New LGTM Enterprise Controller

Reviewer: ```suggestion
### Creating a new LGTM Enterprise control pool
```

> A new LGTM Enterprise controller machine can be created using the `controller.py` template.

Reviewer: ```suggestion
You can create a new LGTM Enterprise control pool machine by using the `controller.py` template.
```

```console
gcloud deployment-manager deployments create --project "$project" --template=controller.py "$deployment" \
	--properties "zone:$zone,administrator-email:email@example.com,administrator-password:MySuperSecretPassword"
```

> The `--properties` flag can be used to provide a variety of options to customize your deployment:

Reviewer: I didn't realise at first that you could use this command pretty much as is, and that the values below were only necessary if you don't like the defaults (which I should probably document in the PDF).
```suggestion
You can use the above command by copying and pasting it, and editing the values for `administrator-email` and `administrator-password`. If you want to customize your deployment, you can add more settings to the `--properties` flag:
```

* `zone` - The Google Cloud zone to create resources in.

> * `virtual-machine-size` - The type of virtual machine to use.

Reviewer: I think we need to give some information here (or refer them to the _LGTM Enterprise Installation and Upgrade Guide_ and give some details there - URL: https://help.semmle.com/lgtm-enterprise/ops/lgtm-enterprise-LATEST-installation-guide.pdf)

Author: You mean the system requirements document?

Reviewer: The Sys Requirements doc doesn't give any specific advice for image-based deployments. I was thinking more of adding information about which machine to choose like we did for AMI installs. See Appendix B of the Install Guide: https://help.semmle.com/lgtm-enterprise/ops/lgtm-enterprise-LATEST-installation-guide.pdf

Though maybe a lot more lightweight than that information.

Author: Thanks, I'll update the link. I thought the CPU and memory requirements might be enough to allow people to select a machine type.

* `data-disk-size-gb` - The data disk size for your LGTM Enterprise instance in gigabytes.
* `administrator-email` - The email address for the initial LGTM Enterprise administrator account.

> * `administrator-password` - The password for the initial LGTM Enterprise administrator account. It is recommended that you change this after the instance has booted.

Reviewer: ```suggestion
* `administrator-password` - The password for the initial LGTM Enterprise administrator account. You should change this when you log in to LGTM Enterprise.
```

> * `general-workers` - The number of general workers to run on the controller.

Reviewer: ```suggestion
* `general-workers` - The number of general workers to run on the control pool host machine.
```

> * `on-demand-workers` - The number of on-demand workers to run on the controller.

Reviewer: ```suggestion
* `on-demand-workers` - The number of on-demand workers to run on the control pool host machine.
```

> * `query-workers` - The number of query workers to run on the controller.

Reviewer: ```suggestion
* `query-workers` - The number of query workers to run on the control pool host machine.
```

* `worker-environment` - A JSON dictionary of environment variables to use for the workers.

> * `manifest-password` - A password used to encrypt the LGTM manifest. If not specified a password will be generated and stored in `/data/lgtm-releases/.manifest-password`.

Reviewer: ```suggestion
* `manifest-password` - A password used to encrypt the LGTM manifest. If you don't specify one a password will be generated and stored in `/data/lgtm-releases/.manifest-password`.
```

> ### Upgrading an Existing LGTM Enterprise Controller

Reviewer: ```suggestion
### Upgrading the LGTM Enterprise control pool
```

> When upgrading, it is important that you use the same properties as when you first deployed LGTM. You can see the properties you used previously by running:

Reviewer: ```suggestion
When upgrading, you must use the same properties you used when you first deployed LGTM. You can see the properties you used previously by running:
```

```
gcloud deployment-manager manifests describe --project "$project" --deployment "$deployment" --format "value(config.content)" "$(gcloud deployment-manager deployments list --project "$project" --filter "name=$deployment" --format "value(manifest)")"
```

> To update an existing LGTM Enterprise controller the old virtual machine and OS disk must be removed.

Reviewer: ```suggestion
To update an existing LGTM Enterprise control pool host the old virtual machine and OS disk must be removed.
```

```
gcloud compute instances delete --project "$project" --zone "$zone" "$deployment" --keep-disks data
```

Next the deployment must be removed, abandoning, but not removing the old resources.
```
gcloud deployment-manager deployments delete --project "$project" --delete-policy abandon "$deployment"
```

> The controller can then be upgraded by re-deploying the template. If you specified any properties when the instance was first created, provide them again here.

Reviewer: ```suggestion
You can now upgrade the control pool by redeploying the template. If you specified any properties when the instance was first created, provide them again here.
```

```console
gcloud deployment-manager deployments create --project "$project" --template controller.py --properties zone:$zone "$deployment"
```

## Workers
Workers are managed as instance groups. All workers in an instance group are provisioned identically and groups can be scaled up and down depending on your needs. Decide a name for your worker group, or if you are upgrading an existing group use the same name you used to create the group.

> 

Reviewer: Add:

Set the `worker_credentials` value by copying and pasting from the **Infrastructure** > **Worker management** tab of the LGTM administration interface.

```console
worker_group="workers"
worker_credentials='worker-credentials: "H4sIAAAAAAAAAI...=="'
```

> ### Creating a New LGTM Enterprise Worker Group

Reviewer: ```suggestion
### Creating a new LGTM Enterprise worker group
```

> A new LGTM Enterprise worker group can be created using the `worker.py` template.

Reviewer: ```suggestion
You can create a new LGTM Enterprise worker group by using the `worker.py` template.
```

```console
gcloud deployment-manager deployments create --project "$project" --template=worker.py "$deployment-$worker_group" \
	--properties "zone:$zone,controller-deployment-name:$deployment,$worker_credentials"
```

> The `--properties` flag can be used to provide a variety of options to customize your deployment:

Reviewer: ```suggestion
You can use the above command by copying and pasting it. If you want to customize your deployment, you can add more settings to the `--properties` flag:
```

> * `controller-deployment-name` - The deployment name for the LGTM Enterprise controller deployment that this worker group should connect to.

Reviewer: ```suggestion
* `controller-deployment-name` - The deployment name for the LGTM Enterprise control pool that this worker group should connect to.
```

> * `worker-credentials` - This should be copied from the administrator interface of the LGTM Enterprise controller.

Reviewer: ```suggestion
* `worker-credentials` - Copy the value from the **Infrastructure** > **Worker management** tab of the LGTM administration interface.
```

* `zone` - The Google Cloud zone to create resources in.

> * `virtual-machine-size` - The type of virtual machine to use.

Reviewer: As for the control pool host, I think we need to give some information here

* `copies` - The number of copies of this worker to create.
* `general-workers` - The number of general workers to run on each machine in the group.
* `on-demand-workers` - The number of on-demand workers to run on each machine in the group.
* `query-workers` - The number of query workers to run on each machine in the group.
* `worker-environment` - A dictionary of environment variables to use for the workers.

> * `manifest-password` - A password used to encrypt the LGTM manifest. If not specified a password will be generated and stored in `/data/lgtm-releases/.manifest-password`.

Reviewer: ```suggestion
* `manifest-password` - A password used to encrypt the LGTM manifest. If you don't specify one a password will be generated and stored in `/data/lgtm-releases/.manifest-password`.
```

> ### Upgrading an Existing LGTM Enterprise Worker Group

Reviewer: ```suggestion
### Upgrading an LGTM Enterprise worker group
```

> When upgrading, you likely want to use the same properties as when you first deployed LGTM. You can see the properties you used previously by running:

Reviewer: ```suggestion
When upgrading, you will probably want to use the same properties as when you first deployed LGTM. You can see the properties you used previously by running:
```

```
gcloud deployment-manager manifests describe --project "$project" --deployment "$deployment-$worker_group" --format "value(config.content)" "$(gcloud deployment-manager deployments list --project "$project" --filter "name=$deployment-$worker_group" --format "value(manifest)")"
```

> The controller can then be upgraded by re-deploying the template. If you specified any properties when the instance was first created, provide them again here.

Reviewer: ```suggestion
You can now upgrade the worker group by redeploying the template. If you specified any properties when the instance was first created, provide them again here.
```
